Timeline history graph - react to click to display entity more info

k3a opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
The timeline history graph card is great, however it lacks details.
Sometimes the intervals between changes are short (as in door sensors), and sometimes the sensor is unavailable for a short period of time, which is not very obvious from the graph.

Therefore, I enhanced the chart by adding a small inline plugin to handle clicking on a timeline element and firing  "hass-more-info" to display the More info dialog of the displayed entity.

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

k3a committed:
Handle clicking on a timeline history graph row to display more info for the displayed entity

bramkragten commented on src/components/chart/ha-chart-base.ts:
```diff
@@ -288,6 +329,25 @@ export class HaChartBase extends LitElement {
         },
       },
+      {
+        id: "labelClickHandler",
```
This should not be in `ha-chart-base`, use the `plugin` property to add a plugin to the chart from just the place you need it.

See for example:

https://github.com/home-assistant/frontend/blob/dev/src/panels/lovelace/cards/energy/hui-energy-devices-graph-card.ts#L156-L166

k3a commented:
Thanks for the suggestion! I have updated it to match the way `hui-energy-devices-graph-card` does it (clicking on a history bar element instead of a label). It is probably batter to have it consistent and I tested that it works fine on mobile too. Tapping opens More info, long tap keeps tooltip only.

It is now implemented in `state-history-chart-timeline` which should be the best spot.
The commit is much smaller too.

bramkragten commented:
We should prevent the action when the graph is the more info itself, as it will give a weird behaviour of opening the more info again.

k3a commented:
I have added `inMoreInfo` (default False) property to `state-history-chart-timeline` and `state-history-charts`.
It is quite specific property for "is in more info dialog" which I don't like that much but this is really the only place where this more-info event should not be fired.

I was considering adding `clickForMoreInfo` (default True) at first. That is more universal and nicer but having it as a property of `state-history-charts` would kind of imply that it applies for `state-history-chart-line` too.
But having more-info handler on `state-history-chart-line` doesn't seem useful to me because More info dialog won't provide more info than that line chart, except maybe the button to access more history.

If you think more-info handler could be useful for the line history graph too, I can rework it into `clickForMoreInfo` property.

bramkragten commented on src/dialogs/more-info/ha-more-info-history.ts:
```diff
@@ -99,6 +99,7 @@ export class MoreInfoHistory extends LitElement {
                 .historyData=${this._stateHistory}
                 .isLoadingData=${!this._stateHistory}
                 .showNames=${false}
+                .inMoreInfo=${true}
```
```suggestion
                inMoreInfo
```

bramkragten committed:
Update src/dialogs/more-info/ha-more-info-history.ts

bramkragten reviewed (approved):
Thanks!

k3a commented:
Actually please take a look at alternative PR #18036 which supports both chart types. I think it makes more sense unless you plan to have some other clicking function on line history state chart in the future.
